Add initial draft of sbom quality section (#247)
chainguard-dev/edu: content/software-security/what-makes-a-good-sbom.md

---
title: "What Makes a Good SBOM?"
description: "An explanation of what makes a good SBOM"
lead: "A guide to SBOM quality"
type: "article"
date: 2022-08-04T15:21:01+02:00
lastmod: 2022-08-04T15:21:01+02:00
contributors: ["John Speed Meyers"]
draft: false
images: []
menu:
  docs:
    parent: "software-security"
weight: 10
toc: true
---

A [software bill of materials](https://edu.chainguard.dev/software-security/glossary/#sbom), or an SBOM (pronounced s-bomb), is a formal record of the components contained in a piece of software. It is analogous to an ingredients list for a recipe. And it has become recognized as one of the key building blocks of software supply chain security. Proponents rightfully point out that organizations can't secure their software if they don't know what's inside their software.

> As awareness and adoption of SBOM has grown, there has been a gradual ackowledgement that [not all SBOMs are created equal](https://www.chainguard.dev/unchained/not-all-sboms-are-created-equal), some are more or less useful, depending on the goals of the SBOM user and the contents of the SBOM. This guides exist to provide some guidance on evaluating the quality of an SBOM, suggesting common use cases and the data fields that support these use cases and open source SBOM quality tools.

Reviewer: ```suggestion
As awareness and adoption of SBOM has grown, there has been a gradual ackowledgement that [not all SBOMs are created equal](https://www.chainguard.dev/unchained/not-all-sboms-are-created-equal), some are more or less useful, depending on the goals of the SBOM user and the contents of the SBOM. This guide exists to provide some guidance on evaluating the quality of an SBOM, suggesting common use cases and the data fields that support these use cases and open source SBOM quality tools.
```

## Basic SBOM Use Cases and Reqiured SBOM Data

**Identify Vulnerable Components**: SBOMs can help organizations and individuals know the unfixed vulnerabilities in their software. By providing an inventory of components, this allows a software maintainer to check whether the versions of their component are associated with any known vulnerabilities. The software maintainer can then update or patch any components with known vulnerabilities.

> To enable this use case, SBOMs must contain information, at a minimum, about the component name and version. Because there are many different open source software package ecosystems, it is also advantageous to include information about the package ecosystem from which a component originates. This reduces the chances of a false positive when identifying potential vulnerabilities. A [package URL](https://github.com/package-url/purl-spec) (or purl) provides this ecosystem information.

Reviewer 2: In addition to the software identifier, maybe we should also list having a checksum as an essential quality feature. We could even add another use case that talks about checking software integrity.

Author: I like it. I will add a third use case about software integrity.

Author: Added :)

> 

Author: Ah, I wanted to keep this first draft simple. Let me add this word about transient dependencies.

> 

Author: Added :)

**Identifying Licenses**: SBOMs can also help organizations and individuals use open source software consistent with the licensing terms of all components. By identifying all components and associated licenses, software teams can understand the legal implications of any decision related to incorporating a particular component.

To enable license analysis, an SBOM must contain information about the license or licenses associated with all components.

Note: There are [other SBOM use cases](https://www.atlanticcouncil.org/in-depth-research-reports/issue-brief/the-cases-for-using-sboms/), such as mapping broader ecosystem risks, but this guide currently focuses on the arguably two most well-established SBOM use cases.

> 

Author: Sounds good. This is helpful. I'll include this in the licensing section, for now.

> 

Author: Included :)

## Measuring SBOM Quality

> The tools for measuring SBOM quality--like the overall concept of SBOM quality itself--are nascent.

Reviewer: I think we are generally doing em-dashes if that's all right

```suggestion
The tools for measuring SBOM quality — like the overall concept of SBOM quality itself — are nascent.
```

> The tools for measuring SBOM quality--like the overall concept of SBOM quality itself--are nascent.

Reviewer: I think we are generally doing em-dashes if that's all right. Also I would intro the next two concepts here. 

```suggestion
The tools for measuring SBOM quality — like the overall concept of SBOM quality itself — are nascent. There are currently two tools that are worth watching to evaluate SBOMs.
```

Author: Added in the revised commits.

[SBOM Scorecard](https://github.com/eBay/sbom-scorecard) analyzes both major SBOM formats and returns a composite quality score indicating the extent to which an SBOM possesses key fields, including whether the components in the SBOM contain a purl and whether there are licenses associated with each component.

> [NTIA Conformance Checker](https://github.com/spdx/ntia-conformance-checker) analyzes whether [SPDX](https://spdx.dev/) SBOM documents posess the data fields associated with the so-called "NTIA minimum elements." [The U.S. National Telecommunications and Information Administration "minimum elements"](https://www.ntia.doc.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) include the data fields deemed essential for basic SBOM use cases, including the identification of vulnerable components and the identification of component licenses.

Reviewer: ```suggestion
[NTIA Conformance Checker](https://github.com/spdx/ntia-conformance-checker) analyzes whether [SPDX](https://spdx.dev/) SBOM documents possess the data fields associated with the so-called "NTIA minimum elements." [The U.S. National Telecommunications and Information Administration "minimum elements"](https://www.ntia.doc.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) include the data fields deemed essential for basic SBOM use cases, including the identification of vulnerable components and the identification of component licenses.
```

> 

Reviewer: I think it would be valuable to include a little "learn more" section so people can find out more information. Maybe pointing to some of the policy work currently in flight?

Author: Added! Though I reference a couple chainguard blog posts, not policy work.

## Measuring SBOM Quality

The tools for measuring SBOM quality — like the overall concept of SBOM quality itself — are nascent. There are currently two tools that are worth watching to evaluate SBOMs.

[SBOM Scorecard](https://github.com/eBay/sbom-scorecard) analyzes both major SBOM formats and returns a composite quality score indicating the extent to which an SBOM possesses key fields, including whether the components in the SBOM contain a purl and whether there are licenses associated with each component.

[NTIA Conformance Checker](https://github.com/spdx/ntia-conformance-checker) analyzes whether [SPDX](https://spdx.dev/) SBOM documents possess the data fields associated with the so-called "NTIA minimum elements." [The U.S. National Telecommunications and Information Administration "minimum elements"](https://www.ntia.doc.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) include the data fields deemed essential for basic SBOM use cases, including the identification of vulnerable components and the identification of component licenses.

## Learn More

> Check out Chainguard's blog post on ["Are SBOMs Any Good?"](https://www.chainguard.dev/unchained/are-sboms-any-good-preliminary-measurement-of-the-quality-of-open-source-project-sboms) to see an application of these SBOM quality tools to a dataset of open source project SBOMs. You can also learn about the complications for SBOM quality created by ["software dark matter."](https://www.chainguard.dev/unchained/software-dark-matter-is-the-enemy-of-software-transparency)

Reviewer: ```suggestion
Check out Chainguard's blog post on ["Are SBOMs Any Good?"](https://www.chainguard.dev/unchained/are-sboms-any-good-preliminary-measurement-of-the-quality-of-open-source-project-sboms) to see an application of these SBOM quality tools to a dataset of open source project SBOMs. You can also learn about the complications for SBOM quality created by ["software dark matter."](https://www.chainguard.dev/unchained/software-dark-matter-is-the-enemy-of-software-transparency).
```
